pkg/network/ebpf/c/protocol-classification-helpers.h

```diff
@@ -0,0 +1,225 @@
+static __always_inline bool has_sequence_seen_before(conn_tuple_t *tup, skb_info_t *skb_info) {
+    if (!skb_info || !skb_info->tcp_seq) {
+        return false;
+    }
+
+    u32 *tcp_seq = bpf_map_lookup_elem(&connection_states, tup);
+
+    // check if we've seen this TCP segment before. this can happen in the
+    // context of localhost traffic where the same TCP segment can be seen
+    // multiple times coming in and out from different interfaces
+    if (tcp_seq != NULL && *tcp_seq == skb_info->tcp_seq) {
+        return true;
+    }
+
+    bpf_map_update_elem(&connection_states, tup, &skb_info->tcp_seq, BPF_ANY);
+    return false;
+}
+
+// A shared implementation for the runtime & prebuilt socket filter that classifies the protocols of the connections.
+static __always_inline void protocol_classifier_entrypoint(struct __sk_buff *skb) {
+    skb_info_t skb_info = {0};
+    conn_tuple_t tup = {0};
+
+    // Exporting the conn tuple from the skb, alongside couple of relevant fields from the skb.
+    if (!read_conn_tuple_skb(skb, &skb_info, &tup)) {
+        return;
+    }
+
+    // We process a non empty TCP packets, rather than that - we skip the packet.
+    if (!should_process_packet(skb, &skb_info, &tup)) {
+        return;
+    }
+
+    // Making sure we've not processed the same tcp segment, which can happen when a single packet travels different
+    // interfaces.
+    if (has_sequence_seen_before(&tup, &skb_info)) {
+        return;
+    }
+
+    struct sock **protocol_key_ptr = bpf_map_lookup_elem(&conn_tuple_to_socket_map, &tup);
+    if (protocol_key_ptr == NULL) {
+        log_debug("[protocol_classifier_entrypoint]: protocol_key_ptr is null, aborting\n");
+        return;
+    }
+    struct sock *protocol_key = *protocol_key_ptr;
+    protocol_t *protocol = bpf_map_lookup_elem(&connection_protocol, &protocol_key);
+    protocol_t local_protocol = PROTOCOL_UNCLASSIFIED;
+    if (protocol != NULL) {
+        local_protocol = *protocol;
+    }
+
+    // If we've already identified the protocol of the socket, no need to read the buffer and try to classify it.
+    if (local_protocol != PROTOCOL_UNCLASSIFIED && local_protocol != PROTOCOL_UNKNOWN) {
+        return;
+    }
+
+    char request_fragment[CLASSIFICATION_MAX_BUFFER];
+    bpf_memset(request_fragment, 0, sizeof(request_fragment));
+    read_into_buffer_for_classification((char *)request_fragment, skb, &skb_info);
+
+    protocol_t original_protocol = local_protocol;
+    classify_protocol(&local_protocol, request_fragment, sizeof(request_fragment));
+
+    log_debug("[protocol_classifier_entrypoint]: Calling protocol: %d\n", local_protocol);
+    // If there has been a change in the classification, save the new protocol.
+    if (original_protocol != local_protocol) {
+        bpf_map_update_with_telemetry(connection_protocol, &protocol_key, &local_protocol, BPF_ANY);
```

Comment:
Do we need to delete the entries somewhere?

Reply:
I think it will be problematic, and I believe we should move to LRU maps (and set the min kernel as 4.10)
WDYT?

Reply:
The problem I can think happening with that is as follows. The address of `struct sock` is not guaranteed to be unique. Infact, it may be reused quiet often due to how the kernel caches objects in its heap. If we don't delete the entry before the lifecycle of the `struct sock` object is complete, we may end up classifying a connection with an old value of `protocol`, if the `struct sock` object gets reused.

Maybe we can add a hook on a function like `security_sk_alloc` which is called everytime a new struct sock is created, and delete the entry there?

Reply:
wouldn't it make more sense to you `security_sk_free` instead?

Reply:
Yes that is fine too imo